x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java

```diff
@@ -0,0 +1,260 @@
+        );
+    }
+
+    /**
+     * Expects
+     * LimitExec[500[INTEGER]]
+     * \_AggregateExec[[],[COUNT([2a][KEYWORD]) AS c],FINAL,null]
+     *   \_ExchangeExec[[count{r}#24, seen{r}#25],true]
+     *     \_EsStatsQueryExec[test], stats[Stat[name=*, type=COUNT, query=null]]], query[{"esql_single_value":{"field":"emp_no","next":{"range":{"emp_no":{"lt":10050,"boost":1.0}}}}}][count{r}#40, seen{r}#41], limit[],
+     */
+    public void testCountAllWithEval() {
+        var plan = plan("""
+              from test | eval s = salary | rename s as sr | eval hidden_s = sr | rename emp_no as e | where e < 10050
+            | stats c = count(*)
+            """);
+        var stat = queryStatsFor(plan);
+        assertThat(stat.type(), is(StatsType.COUNT));
+        assertThat(stat.query(), is(nullValue()));
+    }
+
+
+    /**
+     * Expects
+     * LimitExec[500[INTEGER]]
+     * \_AggregateExec[[],[COUNT([2a][KEYWORD]) AS c],FINAL,null]
+     *   \_ExchangeExec[[count{r}#14, seen{r}#15],true]
+     *     \_EsStatsQueryExec[test], stats[Stat[name=*, type=COUNT, query=null]]],
+     *     query[{"esql_single_value":{"field":"emp_no","next":{"range":{"emp_no":{"gt":10040,"boost":1.0}}}}}][count{r}#30, seen{r}#31], limit[],
+     */
+    public void testCountAllWithFilter() {
+        var plan = plan("from test | where emp_no > 10040 | stats c = count(*)");
+        var stat = queryStatsFor(plan);
+        assertThat(stat.type(), is(StatsType.COUNT));
+        assertThat(stat.query(), is(nullValue()));
+    }
+
+    /**
+     * Expects
+     * LimitExec[500[INTEGER]]
+     * \_AggregateExec[[],[COUNT(emp_no{f}#5) AS c],FINAL,null]
+     *   \_ExchangeExec[[count{r}#15, seen{r}#16],true]
+     *     \_EsStatsQueryExec[test], stats[Stat[name=emp_no, type=COUNT, query={
+     *   "exists" : {
+     *     "field" : "emp_no",
+     *     "boost" : 1.0
+     *   }
+     * }]]], query[{"esql_single_value":{"field":"emp_no","next":{"range":{"emp_no":{"gt":10040,"boost":1.0}}}}}][count{r}#31, seen{r}#32], limit[],
+     */
+    public void testCountFieldWithFilter() {
+        var plan = plan("from test | where emp_no > 10040 | stats c = count(emp_no)");
+        var stat = queryStatsFor(plan);
+        assertThat(stat.type(), is(StatsType.COUNT));
+        assertThat(stat.query(), is(QueryBuilders.existsQuery("emp_no")));
+    }
+
+    /**
+     * Expects - for now
+     * LimitExec[500[INTEGER]]
+     * \_AggregateExec[[],[COUNT(hidden_s{r}#8) AS c],FINAL,null]
+     *   \_ExchangeExec[[count{r}#25, seen{r}#26],true]
+     *     \_AggregateExec[[],[COUNT(hidden_s{r}#8) AS c],PARTIAL,8]
+     *       \_EvalExec[[salary{f}#20 AS s, s{r}#3 AS hidden_s]]
+     *         \_FieldExtractExec[salary{f}#20]
+     *           \_EsQueryExec[test], query[{"esql_single_value":{"field":"emp_no","next":{"range":{"emp_no":{"lt":10050,"boost":1.0}}}}}][_doc{f}#42], limit[], sort[] estimatedRowSize[16]
+     */
+    public void testCountFieldWithEval() {
```

Comment:
The eval trips the optimization - I'll work on a separate PR that fixes that in the Logical planner.

Reply:
I've added a TODO.